tests/test_annotation.py

```diff
@@ -288,66 +287,84 @@ def test_annotate_mutations(self):
+        self, restart_name, basic_recipe, helper_functions, tmp_path
+    ):
         # Test the ability of SLiM to load our files after annotation.
-        for ts, basename in self.get_slim_restarts(no_op=True):
-            tables = ts.tables
-            metadata = [m.metadata for m in tables.mutations]
-            has_nucleotides = tables.metadata['SLiM']['nucleotide_based']
-            if has_nucleotides:
-                nucs = [random.choice([0, 1, 2, 3]) for _ in metadata]
-                refseq = "".join(random.choices(pyslim.NUCLEOTIDES,
-                                                k = int(ts.sequence_length)))
-                for n, md in zip(nucs, metadata):
-                    for m in md['mutation_list']:
-                        m["nucleotide"] = n
-            else:
-                refseq = None
-            for md in metadata:
+        ts = basic_recipe["ts"]
+        tables = ts.tables
+        metadata = [m.metadata for m in tables.mutations]
+        has_nucleotides = tables.metadata['SLiM']['nucleotide_based']
+        if has_nucleotides:
+            nucs = [random.choice([0, 1, 2, 3]) for _ in metadata]
+            refseq = "".join(random.choices(pyslim.NUCLEOTIDES,
+                                            k = int(ts.sequence_length)))
+            for n, md in zip(nucs, metadata):
                 for m in md['mutation_list']:
-                    m["selection_coeff"] = random.random()
-            ms = tables.mutations.metadata_schema
-            tables.mutations.packset_metadata(
-                    [ms.validate_and_encode_row(r) for r in metadata])
-            in_ts = pyslim.load_tables(tables, reference_sequence=refseq)
-            # put it through SLiM (which just reads in and writes out)
-            out_ts = self.run_slim_restart(in_ts, basename)
-            # check for equality, in everything but the last provenance
-            self.verify_slim_restart_equality(in_ts, out_ts)
+                    m["nucleotide"] = n
+        else:
+            refseq = None
+        for md in metadata:
+            for m in md['mutation_list']:
+                m["selection_coeff"] = random.random()
+        ms = tables.mutations.metadata_schema
+        tables.mutations.packset_metadata(
+                [ms.validate_and_encode_row(r) for r in metadata])
+        in_ts = pyslim.load_tables(tables, reference_sequence=refseq)
+        # put it through SLiM (which just reads in and writes out)
+        out_ts = helper_functions.run_slim_restart(in_ts, restart_name, tmp_path)
+        # check for equality, in everything but the last provenance
+        self.verify_slim_restart_equality(in_ts, out_ts)
 
 
 class TestReload(tests.PyslimTestCase):
     '''
     Tests for basic things related to reloading with SLiM
     '''
-
-    def test_load_without_provenance(self):
+    @pytest.mark.parametrize(
+        'restart_name, basic_recipe', restarted_recipe_eq("no_op"), indirect=["basic_recipe"])
```

Comment:
So, what's `indirect` do? Makes it cached? [The docs](https://docs.pytest.org/en/stable/reference.html#pytest.python.Metafunc.parametrize) are kinda cryptic...

Reply:
I think it feeds the parameter value through the fixture function. So passing `("restart_WF.slim", "recipe_WF.slim")` to `"restart_name, basic_recipe", ..., indirect=["basic_recipe"]` will call `basic_recipe("recipe_WF.slim")` and return that as the `basic_recipe` value in the test, whereas the `restart_name` parameter in the test function just contains the plain value `"restart_WF.slim"`

Reply:
oh!! gotcha.